internal/dot/graph.go

```diff
@@ -20,53 +20,183 @@
 
 package dot
 
-import "fmt"
+import (
+	"fmt"
+	"reflect"
+)
 
 // Ctor encodes a constructor provided to the container for the DOT graph.
 type Ctor struct {
-	Name    string
-	Package string
-	File    string
-	Line    int
-	Params  []*Node
-	Results []*Node
+	Name        string
+	Package     string
+	File        string
+	Line        int
+	Params      []*Param
+	GroupParams []*Group
+	Results     []*Result
+}
+
+// Param is a parameter node in the graph.
+type Param struct {
+	*Node
+	Optional bool
+}
+
+// Result is a result node in the graph.
+type Result struct {
+	*Node
+
+	// GroupIndex is added to differenciate grouped values from one another.
+	// Since grouped values have the same type and group, their Node / string
+	// representations are the same so we need indices to uniquely identify
+	// the values.
+	GroupIndex int
+}
+
+// Group is a group node in the graph.
+type Group struct {
+	// Type is the type of values in the group.
+	Type    reflect.Type
+	Group   string
+	Results []*Result
```

Comment:
nit: Add a comment explaining that these are references to the result nodes provided by *possibly different* constructors.